### DISCUSSION TITLE
Tuning AWS Rust SDK HTTP request settings for latency-aware Amazon DynamoDB applications

### DISCUSSION THREAD

**Discussion by ymwjbxxq** [awslabs/aws-sdk-rust#557] [Q&A] [answered]
Hello AWS Team,

I am wondering if you can share some insights like this article:
https://aws.amazon.com/blogs/database/tuning-aws-java-sdk-http-request-settings-for-latency-aware-amazon-dynamodb-applications/

I noticed, for example, that the Latency to DynamoDB is usually under 20ms but sometimes goes up to 200ms on the GetItem(pk and sk), and I would like to put a limit and make retries.

I am not sure that the retry is even considered because it throws 2ms after the first maximum duration.

For example, the timeoutAPI, if I set it to 30ms, I get a Duration: 32.83 ms:
`SdkError("request has timed out: API call (single attempt) timeout occurred after 30ms")`

This is the code:

```
let timeout_config = timeout::Config::new()
        .with_api_timeouts(
            timeout::Api::new()
                // This timeout acts at the "Request to a service" level. When the SDK makes a request to a
                // service, that "request" can contain several HTTP requests. This way, you can retry
                // failures that are likely spurious, or refresh credentials.
                .with_call_timeout(TriState::Set(Duration::from_millis(50)))
                // This timeout acts at the "HTTP request" level and sets a separate timeout for each
                // HTTP request made as part of a "service request."
                .with_call_attempt_timeout(TriState::Set(Duration::from_millis(50))),
        )
        .with_http_timeouts(
            timeout::Http::new()
                // A limit on the amount of time an application takes to attempt to read the first byte over
                // an established, open connection after a write request.
                // Also known as the "time to first byte" timeout.
                .with_read_timeout(TriState::Set(Duration::from_millis(30)))
                // A time limit for completing the connect-handshake. The time starts when
                // making an initial connect attempt on a socket.
                .with_connect_timeout(TriState::Set(Duration::from_millis(30))),
        );
    let config = aws_config::from_env()
        .retry_config(RetryConfig::new().with_max_attempts(10))
        .timeout_config(timeout_config)
        .load()
        .await;

    let dynamodb_client = aws_sdk_dynamodb::Client::new(&config);

```

Could you please help me to understand how to set the parameters?

I would like to achieve
“Reduce the request timeout settings: Tune the client SDK parameters requestTimeOut and clientExecutionTimeout to timeout and fail much faster (for example, after 50 milliseconds). This causes the client to abandon high latency requests after the specified time period and then send a second request that usually completes much faster than the first”

Is it possible to do it with Rust?

Thanks
Dan

**Comment by ymwjbxxq** [comment]
Hello @jdisanti and @rcoh 

because the naming is different, could you please help me to write down the same configuration based on this article
https://aws.amazon.com/blogs/database/tuning-aws-java-sdk-http-request-settings-for-latency-aware-amazon-dynamodb-applications/

I think the comparison is:
with_api_timeouts.with_call_timeout -> Java ClientExecutionTimeout
with_api_timeouts.with_call_attempt_timeout -> Java RequestTimeout

I am not sure about the following:
with_http_timeouts.with_read_timeout -> Java SocketTimeout
with_http_timeouts.with_connect_timeout -> Java ConnectionTimeout

Do you mind when you have time to make a comparison officially?

**Reply by Velfi** [reply]
> Do you mind when you have time to make a comparison officially?

@ymwjbxxq Your comparisons are correct 🙂

**Comment by Velfi** [comment, marked as answer]
For anyone else curious about [this question](https://github.com/awslabs/aws-sdk-rust/discussions/557#discussion-4155041), please see these related issues and the discussions on them:
- https://github.com/awslabs/aws-sdk-rust/issues/558
- https://github.com/awslabs/aws-sdk-rust/issues/567

The SDK had a bug where it would not retry timeout errors. We're investigating the best way to make retry behavior more configurable.
